MagneticField/ParametrizedEngine/test/testParabolicParametrizedMagneticField.cc

```diff
@@ -0,0 +1,127 @@
+  do {
+    if (!(file.read((char*)&px, sizeof(float)) && file.read((char*)&py, sizeof(float)) &&
+          file.read((char*)&pz, sizeof(float)) && file.read((char*)&bx, sizeof(float)) &&
+          file.read((char*)&by, sizeof(float)) && file.read((char*)&bz, sizeof(float))))
+      break;
+
+    const auto gp = GlobalPoint(px, py, pz);
+    if (gp.perp2() > Parameters::max_radius2 || fabs(gp.z()) > Parameters::max_z)
+      continue;
+
+    points.push_back(Vector3f(px, py, pz));
+    referenceB_vec.push_back(GlobalVector(bx, by, bz));
+    count++;
+  } while (count < numberOfPoints);
+
+  const size_t size = points.size();
+  // allocate the input and output host buffer in pinned memory accessible by the Platform devices
+  auto in_host = cms::alpakatools::make_host_buffer<float[], Platform>(size);
+  auto out_host = cms::alpakatools::make_host_buffer<float[], Platform>(size);
+  // fill the input buffers, and the output buffer with zeros
+  for (size_t i = 0; i < size; ++i) {
+    in_host[i] = points[i];
+    out_host[i] = {0.};
+  }
+
+  float resolution = 0.2;
+  float maxdelta = 0.;
+  int fail = 0;
+
+  // run the test on each device
+  for (auto const& device : cms::alpakatools::devices<Platform>()) {
+    auto queue = Queue(device);
+    // allocate input and output buffers on the device
+    auto points_dev = cms::alpakatools::make_device_buffer<float[]>(queue, size);
+    auto deviceB = cms::alpakatools::make_device_buffer<float[]>(queue, size);
+
+    // copy the input data to the device; the size is known from the buffer objects
+    alpaka::memcpy(queue, points_dev, in_host);
+    // fill the output buffer with zeros; the size is known from the buffer objects
+    alpaka::memset(queue, deviceB, 0.);
+
+    auto workDiv = cms::alpakatools::make_workdiv<Acc1D>(1, size);
+    alpaka::exec<Acc1D>(queue, workDiv, MagneticFieldKernel, points_dev.data(), deviceB.data(), size);
+
+    // copy the results from the device to the host
+    alpaka::memcpy(queue, out_host, deviceB);
+
+    // wait for the kernel and the potential copy to complete
+    alpaka::wait(queue);
+
+    // check the results
+    for (uint i=0; i<points.size(); i++) {
+      const auto point = points[i];
+      const auto referenceB = referenceB_vec[i];
+      const auto out_host_element = out_host[i];
+      GlobalVector parametricB(out_host_element[0], out_host_element[1], out_host_element[2]);
+      if ((referenceB - parametricB).mag() > resolution) {
+        ++fail;
+        float delta = (referenceB - parametricB).mag();
+        if (delta > maxdelta)
+          maxdelta = delta;
+        if (fail < 10) {
+          const GlobalPoint gp(point(0), point(1), point(2));
+          cout << " Discrepancy at point  # " << count + 1 << ": " << gp << ", R " << gp.perp() << ", Phi " << gp.phi()
+               << ", delta: " << referenceB - parametricB << " " << delta << endl;
+          cout << " Old: " << referenceB << ", New: " << parametricB << endl;
```

Comment:
"Reference" is a good suggestion; please note that "Approximation" may be confusing because it depends what Reference is. If existing reference files are used, indeed this would be comparing the result of the "standard" MF with the parabolic parametrization. But nothing prevents that at some point we create a reference file from the parametrization itself so that the comparison is a real 1:1 test - for this reason I would rather suggest "Reference"  and "AlpakaParametrization" or something like this.
[BTW the name of the test is also somewhat misleading since we do have a non-alpaka ParabolicParametrizedMagneticField, and the test name seems to refer to it. If that is not too much of a problem I'd prefer this test file to include "alpaka" in its name.
]

Reply:
Actually, no.

The fact that we use alpaka for portability is kind of an implementation detail, normally we don't include it as part of the class and file names.
The common exceptions are the implementation of the alpaka utilities themselves, and some of the pixel code where we needed to distinguish cuda-based and alpaka-based versions; we plan to rename the latter after removing the old cuda-based implementations.

Also, as Matti pointed out, the code in `MagneticField/ParametrizedEngine/interface/alpaka/ParabolicParametrizedMagneticField.h` is not actually using alpaka.

Reply:
Then I do not see what is the point of this test at all. We already have the subsystem's unit test, which is [MagneticField/Engine/test/testMagneticField.cc](https://github.com/cms-sw/cmssw/blob/master/MagneticField/Engine/test/testMagneticField.cc), called [here](https://github.com/cms-sw/cmssw/blob/master/MagneticField/Engine/test/runTest.sh). My understanding is that this file, which is largely a copy of the above as I suggested to Lorenzo, is required specifically because you also want to test the alpaka implementation, which is not easily integrated in MagneticField/Engine/test/testMagneticField.cc. I am fine with this providing that this is *only* used for the alpaka implementation, we should not replicate existing tests and code.

Reply:
The use of `.../alpaka/...` in the path names is related to the _implementation_ of the build rules.

That implementation is based on the alpaka library; as such it seems reasonable to use `alpaka` as the name of the subdirectory used to separate the alpaka-based and non alpaka-based files.

The name of the classes and functions is related to their purpose, not their implementation; as such it makes more sense to use the word "portable" in the name if anything is even needed. In fact it would prefer to rename the old code to "legacy" and leave the new code unencumbered by adjectives, but I realise that would take a lot of work for no practical gain.

Barring that, we are currently trying to use "SoA" for the new data structures and "Portable" for new the modules.

Back to the code at hand: the parametric magnetic field is _trivial_, implemented using a few very simply functions.
It should be usable at runtime (and even at compile time) taking advantage of its simplicity. For example, going through virtual function indirection is likely to add more overhead than the time it takes to perform the actual computation.

Using runtime values for the parameters instead of compile time constants is going to be slightly slower, but likely not much so.
So, if there will be a _need_ to support different magnetic field values, that's likely the best way forward.
I insist on "need", and not on "it should be done just because it's what is being done somewhere else".

As to where the code of the magnetic field should go an how it should be implemented:
  - if the code needs to be compiled into different symbols for the different backends, it should be wrapped in the `ALPAKA_ACCELERATOR_NAMESPACE` and be put in the `interface/alpaka` and `src/alpaka/` directories;
  - otherwise, if standard c++ features like `constexpr` functions and different library implementations are enough, it should go under `interface/` and `src/`.

On the other hand, the test _has to_ use alpaka to test the different backends, so it needs to go under `test/alpaka/` anyway.

Reply:
> Then I do not see what is the point of this test at all.

Because the current code does not run on gpus, so new code is needed, and a test for the new code should be added ?

Reply:
That's exactly why discussing with @lecriste we agreed on replicating the existing unit test in a new test file rather than trying to integrate this use case into it. 
We do agree that new code is needed to test new functionality, what I am asking is not to name it in a way that it seems to be intended to be used to test the "old" code instead, or as well. It seems to me a very reasonable request, I don't see what is the problem with this.

Reply:
Indeed, I don't have any problems with this request.
I simply explained why we should not include "Alpaka" in the name, and why it's indeed OK to include "Portable" in the name.

Reply:
> Back to the code at hand: the parametric magnetic field is _trivial_, implemented using a few very simply functions. It should be usable at runtime (and even at compile time) taking advantage of its simplicity. For example, going through virtual function indirection is likely to add more overhead than the time it takes to perform the actual computation.

I fully agree with this, but on the other hand, one should not forget the full picture. This is just one of the many different parametrization that we are asked to support - W mass people for example wanted to plug in the best possible model in terms of accuracy regardelss of efficiency, while general simulation and reconstruction is enough with something inbetween, but requires a single model for the whole of CMS. The current infrastructure provides a configurable service to cover these different use cases in an interchangeable way. It's difficult to do the same without relying on a virtual interface. . I just think we have to be sure that "hardcoded" implementations are clearly understood for what they do, and for their limits -for example, this one, besides being valid only in the central region, is tuned to reproduce the field integral, not the actual value of B at a specific point. This is why I am concerned with seeing this discussed as a perfect alternative to the default map.
 
> Using runtime values for the parameters instead of compile time constants is going to be slightly slower, but likely not much so. So, if there will be a _need_ to support different magnetic field values, that's likely the best way forward. I insist on "need", and not on "it should be done just because it's what is being done somewhere else".

I think I am missing the point here. Parameters are fixed for a given working point of the magnet, but have to be adjusted for the different working points. Our magnet may happen to run at B!=3.8T for real, not because this is being done somewhere else in the code :-D

Reply:
> Our magnet may happen to run at B!=3.8T for real, not because this is being done somewhere else in the code :-D

Eh, 15 years of history say that we are not going to take physics data with different values of the magnetic field.

If at some point (e.g. for Run-4 or Run-5) we are forced to move to a different value, we can simply change the parametrisation.

Reply:
You seem to miss that we have been operating our detector several times at reduced currents, even B=0, and even if that did not yield physics data. Myself, I do remember people trying to figure out why their code was going nuts because they were assuming B=3.8T and they were trying to reconstruct B=0 cosmics.

Funny enough, no more than [five days ago](https://github.com/cms-sw/cmssw/issues/45392#issuecomment-2217908641), somebody complained because processing data taken at different currents in the same job (!) did not work.

So yes, please go on with hardcoding values! Somebody else will have to scratch their head when they will have to run on non-physics data and will have no idea of why they get meaningless results.

Reply:
> You seem to miss that we have been operating our detector several times at reduced currents, even B=0, and even if that did not yield physics data. Myself, I do remember people trying to figure out why their code was going nuts because they were assuming B=3.8T and they were trying to reconstruct B=0 cosmics.

I am thoroughly aware of that - as I am aware of the fact that all the collisions data at B != 3.8 T has been scrapped.

So the question is: if we can speed up by 10% our reconstruction code by ignoring the possibility of running at B != 3.8, is it worth it ?
What if we speed it up by a factor 2? Or a factor 10 ?

> Funny enough, no more than [five days ago](https://github.com/cms-sw/cmssw/issues/45392#issuecomment-2217908641), somebody complained because processing data taken at different currents in the same job (!) did not work.

Indeed, and the conclusion was 

"While for academic purpose this may be interesting, can the 370616 just be removed from processing?"

> So yes, please go on with hardcoding values! Somebody else will have to scratch their head when they will have to run on non-physics data and will have no idea of why they get meaningless results.

Sure - or they can learn to preselect what they run on.

Reply:
People has been analyzing data at eg B=0 for several purposes. I don't think we can just assume the data goes to /dev/null when B!=0. 
Besides, in case of  magnet issues like what we have experienced in summer 2012, we may find that we just need to keep what we have right away, without waiting the deployment of a new release. These are the reasons why conditions-dependent MF was made into a requirement (and not by myself)

However, although I enjoy discussing the lessons of history, why don'y we rather focus on a way to choose one among 5 sets of 4 hardcoded parameters depending on run conditions. It should not be rocket science.

Reply:
> Indeed, and the conclusion was
> 
> "While for academic purpose this may be interesting, can the 370616 just be removed from processing?"

Note that the conclusion was that we (actually, dd4hep) cannot support running **in the same job** over several runs taken at (some) of the different working points.
This is very different than not supporting any B other than 3.8T.